Onboarding and settings refactor to React

iamdharmesh opened the pull request:
### Description of the Change
The PR refactors the settings and onboarding to a React-based implementation and deprecates the existing settings panel. The existing settings panel is deprecated but can still be enabled using the `classifai_use_legacy_settings_panel` filter.

![screencapture-classifai-local-wp-admin-tools-php-2024-09-19-16_17_44](https://github.com/user-attachments/assets/8fd905bf-cefa-4349-b9df-321d2220719f)

Closes #502
Closes #449 

### How to test the Change
**Settings:**
1. Go to `Tools` > `ClassifAI`.
1. Ensure the settings page loads correctly and that navigation works as expected.
1. Update feature settings, such as enabling/disabling features or changing/configuring providers, and verify that everything functions as expected.
1. **Existing settings:** Configure the ClassifAI plugin using the `develop` branch, then switch to this PR branch. Confirm that no settings are lost and that all data is correctly reflected in the new settings page.
1. **Use Legacy Settings:** Add the following code to the theme’s functions.php file or plugin: `add_filter( 'classifai_use_legacy_settings_panel', '__return_true' );`. Verify that the old settings page loads and works as expected.

**OnBoarding**
1. Deactivate the ClassifAI plugin and activate it again.
2. You will see a notice to start the ClassifAI setup. Click the "Start Setup" button.
3. Follow the setup process and ensure everything works as expected.
4. **Existing settings:** Visit the setup page from the settings page and verify that it correctly reflects the existing configured features, with the option to modify the settings accordingly.
5. **Use Legacy Settings:** Add the following code to the theme’s functions.php file or plugin: `add_filter( 'classifai_use_legacy_settings_panel', '__return_true' );`. Verify that the old onboarding/setup page loads and works as expected.

**Add additional feature fields**
1. Add a new feature to the plugin by following the guide [here](https://github.com/10up/classifai/blob/enhancement/react-settings/hookdocs/useful-snippets.md#add-a-new-feature).
2. Follow step 4 to add additional feature fields to the settings and validate that everything works as expected.

**Add custom Provider fields**
1. Add a custom provider to any feature by following the guide [here](https://github.com/10up/classifai/blob/enhancement/react-settings/hookdocs/useful-snippets.md#add-custom-provider-to-an-existing-feature).
2. Add the required provider fields to the settings by following step 5 in the guide and validate that everything works as expected.

**General**
Run a full regression test to ensure everything is working as expected.

### Changelog Entry
> Added - New React-based settings and onboarding process.
> Deprecated - The existing settings panel.

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @jeffpaul @dkotter @iamdharmesh @Sidsector9 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests pass.

iamdharmesh committed:
Updated current setting page slug to classifai_old.

iamdharmesh committed:
Initial base react setting page.

iamdharmesh committed:
Created user access component.

iamdharmesh committed:
Improved folder structure and some other improvements.

iamdharmesh committed:
Add data store for state management.

iamdharmesh committed:
Update components as per to utilize data store.

iamdharmesh committed:
Added sample providerSettings and additionalFeature settings filters.

iamdharmesh committed:
Improvements in root component.

iamdharmesh committed:
Add current service and feature releated functions in useSettings hook.

iamdharmesh committed:
Update in settings-wrapper component.

iamdharmesh committed:
Set default service and feature based on tab.

iamdharmesh committed:
Re-structure folders in components folder.

iamdharmesh committed:
Make use of <PluginArea /> to extend provider and additional feature settings.

iamdharmesh committed:
Extend feature settings for comment-moderation feature.

iamdharmesh committed:
Extend open-moderation provider fields.

Sidsector9 commented:
@iamdharmesh below, I've checked the "Editor" checkbox under "Allowed roles". Ideally it should have just re-rendered the area pointed by the arrow, but instead the whole App re-renders. We can consider improving this.

<img width="1827" alt="Screenshot 2024-05-20 at 9 20 17 PM" src="https://github.com/10up/classifai/assets/17757960/9843e388-613b-4f4c-8f5b-4faeee00a376">

faisal-alvi committed:
narrow down the container

iamdharmesh committed:
Fix header links.

faisal-alvi committed:
button full width; min heights to tabs

Sidsector9 commented:
Preventing access of state at the root App may help avoid re-rendering. I am working on and likely to fix this.

Sidsector9 committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

Sidsector9 committed:
improve re-rednering performance

Sidsector9 commented:
Improved areas where re-rendering wasn't necessary. I'm trying to further improve this.

<img width="2902" alt="Screenshot 2024-05-28 at 8 19 06 PM" src="https://github.com/10up/classifai/assets/17757960/12741276-59cc-472c-84bb-bb2ff905f167">

Sidsector9 committed:
further improve performance

Sidsector9 committed:
remove unused deps

iamdharmesh committed:
Move allowed roles to separate component.

iamdharmesh committed:
Design improvements.

iamdharmesh committed:
Changed tab panel with react router dom.

iamdharmesh committed:
Code Improvements

iamdharmesh committed:
Simplified extended code.

iamdharmesh committed:
Improved component re-rendering

iamdharmesh committed:
Added provider fields for the chatGPT, Gemini API and Azure OpenAI

iamdharmesh committed:
Temporary stop E2E tests.

iamdharmesh committed:
Updated update  settings API

iamdharmesh committed:
Added error notice in settings.

iamdharmesh committed:
Removed unwanted hook.

jeffpaul commented:
@iamdharmesh wanted to ensure you're looking at leveraging WP core components for most of this, possibly using https://vajrawp.smalltowndev.com/ for inspiration, and seeing how https://wordpress.org/plugins/block-visibility/ handles their settings experience as another decent approach that uses WP components with some brand styling.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Use context to avoid props-drilling

iamdharmesh committed:
Remove unwanted doc comments.

iamdharmesh committed:
Added provider settings for IBM watson NLU.

iamdharmesh committed:
Updated UI for configured providers.

iamdharmesh committed:
Added additional openai providers.

iamdharmesh committed:
Added some more providers

iamdharmesh committed:
Match component style with admin color.

iamdharmesh committed:
Started Adding additional feature setting fields.

iamdharmesh committed:
Indicate saving settings.

iamdharmesh committed:
Added OpenAI dalle provider settings.

iamdharmesh committed:
Added additional settings for image tags generator feature.

iamdharmesh committed:
Add AWS polly providers.

iamdharmesh committed:
Added Azure Text to Speech provider settings.

iamdharmesh committed:
Added provider settings for the OpenAI TTS

iamdharmesh committed:
Added TTS feature settings.

iamdharmesh committed:
Added Title generation settings.

iamdharmesh committed:
Content resizing settings.

iamdharmesh committed:
Excerpt generation settings.

iamdharmesh committed:
Initial classification settings.

iamdharmesh committed:
Added classification settings.

iamdharmesh committed:
Added moderation settings.

iamdharmesh committed:
Make User Permissions panel toggle persistence.

iamdharmesh committed:
Add ClassifAI registration

Sidsector9 committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

Sidsector9 committed:
Merge branch 'develop' into enhancement/react-settings

Sidsector9 committed:
add fragment between Tooltip and Icon

Sidsector9 committed:
create the onboarding page

Sidsector9 committed:
use hook instead

Sidsector9 commented:
I'd like to share something that I learnt while contributing to this PR.

If we're building reusable Components that can be later used somewhere else, then Contexts should be used very wisely.
Contexts are notorious for restricting Component reusability, in the sense that if a Component accesses context data and is dependent on it, then that Component can only be used unless a Context Provider is used, and that may not always be the case.

A good thumb rule while building reusable Components would be to think of possible long-term use cases.

@10up/open-source-practice I'd like to hear your views on this.

Sidsector9 committed:
add onboarding page

iamdharmesh committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Display notice at feature level.

iamdharmesh committed:
Add personalizer deprecation notice.

iamdharmesh committed:
Hide credentials fields if provider is configured.

iamdharmesh committed:
Fix PHPCS errors.

iamdharmesh committed:
Remove unwanted files.

iamdharmesh committed:
Remove admin component.

iamdharmesh committed:
Refactor onboarding steps.

iamdharmesh committed:
Added onSaveSuccess callback support.

iamdharmesh committed:
Remove layout component.

iamdharmesh committed:
Update finish step designs.

iamdharmesh committed:
Refactor configure feature step.

iamdharmesh committed:
Removed unwanted state from the onboarding process.

iamdharmesh committed:
Update onboarding header navigation.

iamdharmesh committed:
Added button links to last step.

iamdharmesh committed:
Add registration step.

Sidsector9 committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

Sidsector9 committed:
add previewer search control

Sidsector9 committed:
fetch results

Sidsector9 committed:
color code results

Sidsector9 committed:
add context for previewer

Sidsector9 committed:
add notice

Sidsector9 committed:
add close previewer button

Sidsector9 committed:
modify embeddings result data format

Sidsector9 committed:
add support for OpenAIEmbedding

Sidsector9 committed:
add previewer for IBMWatson

Sidsector9 committed:
move provider previewers into separate components

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Adding style to loading spinner.

iamdharmesh committed:
Some design fixes on onboarding.

iamdharmesh committed:
Move notices under header.

iamdharmesh committed:
Add current step stat on configure feature tabs in onboarding.

iamdharmesh committed:
Skip registration step if user is already registered.

Sidsector9 committed:
fix embeddings old previewer

iamdharmesh committed:
remove strict versioning from .nvmrc

iamdharmesh committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

Sidsector9 committed:
fix eslint errors

Sidsector9 committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

Sidsector9 committed:
fix phpcs errors

iamdharmesh committed:
Added filter `classifai_use_legacy_settings_panel` to enable legacy settings panel.

iamdharmesh committed:
Fix setup role in rest API settings endpoint.

iamdharmesh committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Fix setup link in plugin links and notice.

iamdharmesh committed:
Remove unwanted code.

iamdharmesh committed:
Update regenerate embeddings processing code

iamdharmesh committed:
Fix Sass compile warning.

iamdharmesh committed:
Fix default feature settings.

iamdharmesh committed:
Fix IBM watson auth saving

iamdharmesh committed:
Update admin tests.

iamdharmesh committed:
Update tests for the azure open AI excerpt generation.

iamdharmesh committed:
Update E2E tests for google gemini API - excerpt generation.

iamdharmesh committed:
Update E2E tests for openAI ChatGPT

iamdharmesh committed:
Update E2E tests for moderation feature.

iamdharmesh committed:
Update E2E tests for resize content feature.

iamdharmesh committed:
Update E2E tests for speech to text feature.

iamdharmesh committed:
Update E2E tests for text to speech feature.

iamdharmesh committed:
Update tests for the title generation feature.

iamdharmesh committed:
Updated E2E tests for the IBM Watson NLU Classification feature.

iamdharmesh committed:
Update E2E tsets for the classification feature.

iamdharmesh committed:
Update tests for image generation feature.

iamdharmesh committed:
Update E2E tests for the image processing feature.

iamdharmesh committed:
Update E2E tests for PDF read feature.

iamdharmesh committed:
Update E2E tests for the common feature fields.

iamdharmesh committed:
Fix failing test.

iamdharmesh committed:
Enable E2E tests workflow.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Update react render to make it work in WP < 6.2 as well.

iamdharmesh committed:
get post status options using localize scripts.

iamdharmesh committed:
FIx settings in WP minimum env.

iamdharmesh committed:
Header design improvements.

iamdharmesh committed:
Add react settings for the smart 404 feature.

iamdharmesh committed:
Display error notice if EP is not install and active.

iamdharmesh committed:
Update E2E tests for Smart 404 feature.

iamdharmesh committed:
Added command for enable disable features.

iamdharmesh committed:
Improvement in E2E tests and fix E2E tests for WP minimum environment.

iamdharmesh committed:
Fix E2E tests Smart 404 feature.

iamdharmesh committed:
Fix previewer for azure OpenAI embeddings provider.

iamdharmesh committed:
Update add feature and provider documentation for add fields to react-based settings.

iamdharmesh committed:
Update documentation for the keep using legacy settings panel.

iamdharmesh committed:
Added error handling.

iamdharmesh committed:
Remove unwanted files.

iamdharmesh committed:
display previewer only if provider is configured and feature is active.

iamdharmesh committed:
Some slot name updates.

iamdharmesh committed:
Sanitize status field in onboarding settings save.

iamdharmesh committed:
Improved code by moving personalizer notice to separate component.

iamdharmesh committed:
Display notice for the in progress background processing of embedding generation.

iamdharmesh committed:
Use v4.3.3 of deps review.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/classifai into enhancement/react-settings

iamdharmesh committed:
Fix PHPStan error.

Sidsector9 committed:
Merge branch 'develop' of github.com:10up/classifai into develop

Sidsector9 committed:
Merge branch 'enhancement/react-settings' of github.com:10up/classifai into enhancement/react-settings

Sidsector9 committed:
Merge branch 'develop' into enhancement/react-settings

Sidsector9 commented on hookdocs/useful-snippets.md:
```diff
@@ -185,6 +172,87 @@ add_filter(
+		} );
+
+	return (
+		<Fill name="ClassifAIProviderSettings">
```
**Note to self:**

Find out a way to make this easier so that 3rd-party devs don't have to write `Fill`.
Maybe try exporting a component on the `window` object which can be used to inject custom Provider settings.

dkotter committed:
Minor updates to ensure we have return typing on new functions

dkotter committed:
Update most dependencies

dkotter reviewed (commented):
Great work on this @iamdharmesh and @Sidsector9, this looks really nice! I've not gone through all the code line by line but at a high level this all looks good to me.

Couple thoughts:

1. We're introducing a lot of new JS files and doesn't seem any of them have code comments in place explaining what is going on. A lot of those are self explanatory (like all the provider settings ones) but I think some of the others could benefit from either a comment at the top of the file or a comment on the individual exported methods explaining what they're used for, just making it more clear for anyone looking at the code
2. I like how the `User permissions` are broken into their own section from the main settings. Could we collapse that section by default? For me it was open and I had to collapse it myself but I think a default of collapsed would be nice
3. Curious on any discussion that was had around keeping the old settings and just hiding that behind a filter? I almost lean towards just getting rid of all that, so we're not having to maintain a bunch of legacy code
4. @jeffpaul I think the new design looks really good but being such a major change, is this something we'd want to pull in UX resources to give a quick look at (if we haven't already)?

dkotter committed:
Add missing textdomain

iamdharmesh committed:
Keep  User permissions collapsed by default.

iamdharmesh committed:
Add Doc comments in JS files.

iamdharmesh commented:
Thank you very much for the review and for providing feedback, @dkotter.

> We're introducing a lot of new JS files and doesn't seem any of them have code comments in place explaining what is going on. A lot of those are self explanatory (like all the provider settings ones) but I think some of the others could benefit from either a comment at the top of the file or a comment on the individual exported methods explaining what they're used for, just making it more clear for anyone looking at the code

Thanks for this note. Addressed in https://github.com/10up/classifai/pull/771/commits/73150b5d1a2801c959dfa1d4b982938e993eb38c

> I like how the User permissions are broken into their own section from the main settings. Could we collapse that section by default? For me it was open and I had to collapse it myself but I think a default of collapsed would be nice

Agree. Addressed in https://github.com/10up/classifai/pull/771/commits/bc39ea0cc126e8d6904a80b7fb1667395ce9a74f

> Curious on any discussion that was had around keeping the old settings and just hiding that behind a filter? I almost lean towards just getting rid of all that, so we're not having to maintain a bunch of legacy code.

I discussed this with @jeffpaul in our 1:1. This was intended to allow users who have extended or customized our old settings some extra time until we remove these settings in the upcoming major release. However, I am open to completely removing the old setup if that seems like the best path forward.

Thank you.

iamdharmesh committed:
E2E tests update to handle collapsed "User Permissions" panel.

iamdharmesh committed:
E2E: Open "User permissions" panel if closed.

jeffpaul commented:
> I discussed this with @jeffpaul in our 1:1. This was intended to allow users who have extended or customized our old settings some extra time until we remove these settings in the upcoming major release. However, I am open to completely removing the old setup if that seems like the best path forward.

I'm also open to jettisoning that code here to keep things easier to maintain on our end.  Perhaps some note in the dev docs on migrating any extensions of the former settings?

Sidsector9 commented:
### Update

The following is the UI update as per the new design. I'll raise a separate PR.

<img width="773" alt="Screenshot 2024-11-10 at 6 54 55 PM" src="https://github.com/user-attachments/assets/ac9ef6b6-dff4-448f-8e1a-1858ecc1a033">

dkotter commented:
> This was intended to allow users who have extended or customized our old settings some extra time until we remove these settings in the upcoming major release. However, I am open to completely removing the old setup if that seems like the best path forward.

> I'm also open to jettisoning that code here to keep things easier to maintain on our end. Perhaps some note in the dev docs on migrating any extensions of the former settings?

I'd love to just remove it rather than having it clutter up our codebase. That said, as long as we open an issue to track the removal of this in a release or two, that's fine by me.

I do think having docs around how to migrate would be nice, though not a blocker to getting this merged in.

> The following is the UI update as per the new design. I'll raise a separate PR.

@Sidsector9 Sounds like the plan is to make these updates in a separate PR, is that correct? If so, @Sidsector9, @iamdharmesh I'm fine with this getting merged in and we can iterate on things in separate PRs. Note we have a few open PRs that I think will need to be modified to work in the new settings once this gets merged in as well.
